data-processing/entities/definitions/model/model/modeling_joint_roberta.py

```diff
@@ -108,56 +115,56 @@ def forward(
         ):
             pooled_output = self.custom_pooler(sequence_output)
 
-        intent_logits = self.intent_classifier(pooled_output)
-        slot_logits = self.slot_classifier(sequence_output)
+        intent_logits = {}
+        slot_logits = {}
+
+        for classifier_type, intent_classifier in self.intent_classifiers.items():
+            intent_logits[classifier_type] = intent_classifier(pooled_output)
+
+        for classifier_type, slot_classifier in self.slot_classifiers.items():
+            slot_logits[classifier_type] = slot_classifier(sequence_output)
 
         total_loss = 0
+        total_loss_dict = {}
+        
         # 1. Intent Softmax
-        if intent_label_ids is not None:
-            if self.num_intent_labels == 1:
-                intent_loss_fct = nn.MSELoss()
-                intent_loss = intent_loss_fct(
-                    intent_logits.view(-1), intent_label_ids.view(-1)
-                )
-            else:
-                intent_loss_fct = nn.CrossEntropyLoss()
-                intent_loss = intent_loss_fct(
-                    intent_logits.view(-1, self.num_intent_labels),
-                    intent_label_ids.view(-1),
-                )
-            total_loss += intent_loss
+        for data_type in self.tasks:
+            intent_label_id = intent_label_ids[data_type]
+            if intent_label_id is not None:
+                if self.num_intent_labels_dict[data_type] == 1:
+                    intent_loss_fct = nn.MSELoss()
+                    intent_loss = intent_loss_fct(intent_logits[data_type].view(-1), intent_label_id.view(-1))
+                else:
+                    intent_loss_fct = nn.CrossEntropyLoss()
+                    intent_loss = intent_loss_fct(intent_logits[data_type].view(-1, self.num_intent_labels_dict[data_type]), intent_label_id.view(-1))
+                total_loss += intent_loss
+                total_loss_dict["intent_{}".format(data_type)] = intent_loss
```

Comment:
have you heard of f strings (e.g., `f"intent_{data_type}"`)? They are a nifty Python feature that provides a subset of the string formatting functionality. Not necessary to change, just wanted to share.

Reply:
Yes! I love fstrings! This was legacy code from the trainer repo that I just got lazy to change. 😛